File: data/evaluations/2026-09-05-codegen-pilot-01/deterministic-v1-ts-c0d54e8e15702df309af/evaluator-source/src/evaluation/scoring.ts
/** Pure deterministic scoring. Integer half-points avoid intermediate rounding. */
import { equal } from "../lib/io";
export const CATEGORIES = {
  functional_requirements: 30,
  connectivity: 30,
  physical_constraints: 20,
  deliverable_integrity: 20,
};
export type Category = keyof typeof CATEGORIES;
export type Outcome = "pass" | "fail" | "unknown" | "unsupported";
export interface RuleTest {
  id: string;
  category: Category;
  critical: boolean;
  rule: Record<string, any>;
  requirement: string;
  source_criteria: string[];
}
export interface Rules {
  version: string;
  weights: typeof CATEGORIES;
  applicability: Record<Category, boolean>;
  tests: RuleTest[];
  expected_components: any[];
  [key: string]: any;
}
export interface Decision {
  test_id: string;
  outcome: Outcome;
  evidence: string[];
  [key: string]: any;
}
export function validateRules(rules: Rules) {
  if (rules.version !== "1.0.0") throw Error("Unsupported rules version");
  if (!equal(rules.weights, CATEGORIES))
    throw Error("Rules must use the frozen 30/30/20/20 weights");
  if (
    !equal(
      Object.keys(rules.applicability ?? {}).sort(),
      Object.keys(CATEGORIES).sort(),
    ) ||
    Object.values(rules.applicability).some((x) => typeof x !== "boolean")
  )
    throw Error("Every category needs explicit boolean applicability");
  const seen = new Set<string>(),
    counts = Object.fromEntries(Object.keys(CATEGORIES).map((k) => [k, 0]));
  for (const t of rules.tests ?? []) {
    if (typeof t.id !== "string" || !t.id || seen.has(t.id))
      throw Error("Missing or duplicate test ID");
    seen.add(t.id);
    if (!(t.category in CATEGORIES) || typeof t.critical !== "boolean")
      throw Error("Invalid category or critical flag");
    if (
      !t.rule ||
      typeof t.rule !== "object" ||
      Array.isArray(t.rule) ||
      !t.requirement ||
      !t.source_criteria?.length
    )
      throw Error(
        "Every test needs an executable rule and requirement provenance",
      );
    counts[t.category]++;
  }
  if (
    Object.keys(CATEGORIES).some(
      (c) => rules.applicability[c as Category] && !counts[c],
    )
  )
    throw Error("Applicable categories cannot have zero tests");
  return rules;
}
export function score(rules: Rules, outcomes: Decision[]) {
  validateRules(rules);
  const tests = rules.tests;
  if (
    outcomes.length !== tests.length ||
    !equal(
      [...new Set(outcomes.map((o) => o.test_id))].sort(),
      tests.map((t) => t.id).sort(),
    )
  )
    throw Error("Missing, duplicate, or unexpected outcomes");
  for (const o of outcomes) {
    if (!["pass", "fail", "unknown", "unsupported"].includes(o.outcome))
      throw Error("Invalid outcome");
    if (["pass", "fail"].includes(o.outcome) && !o.evidence?.length)
      throw Error("Pass/fail requires evidence");
  }
  const byId = Object.fromEntries(outcomes.map((o) => [o.test_id, o]));
  let points = 0,
    lower = 0,
    upper = 0,
    denominator = 0,
    anyUnknown = false,
    criticalFailure = false;
  const all: Outcome[] = [];
  const categories = Object.entries(CATEGORIES).map(([c, weight]) => {
    const category = c as Category,
      applicable = rules.applicability[category],
      selected = applicable
        ? tests
            .filter((t) => t.category === category)
            .map((t) => byId[t.id].outcome)
        : [];
    const passed = selected.filter((o) => o === "pass").length,
      failed = selected.filter((o) => o === "fail").length,
      unknown = selected.length - passed - failed;
    let value: number | null = null,
      lo: number | null = null,
      hi: number | null = null;
    if (applicable) {
      denominator += weight;
      all.push(...selected);
      if (unknown) {
        anyUnknown = true;
        lo = passed ? 1 : 0;
        hi = failed ? 1 : 2;
      } else {
        value = passed === selected.length ? 2 : passed ? 1 : 0;
        lo = hi = value;
        points += weight * value;
      }
      lower += weight * lo;
      upper += weight * hi;
      criticalFailure ||= tests.some(
        (t) =>
          t.category === category &&
          t.critical &&
          byId[t.id].outcome === "fail",
      );
    }
    return {
      category,
      weight,
      applicable,
      score: value,
      possible_score_min: lo,
      possible_score_max: hi,
      passed,
      failed,
      unknown_or_unsupported: unknown,
    };
  });
  return {
    category_scores: categories,
    total_score:
      denominator && !anyUnknown ? (50 * points) / denominator : null,
    overall_pass: criticalFailure
      ? false
      : anyUnknown || !denominator
        ? null
        : all.every((o) => o === "pass"),
    critical_failure: criticalFailure,
    possible_total_min: denominator ? (50 * lower) / denominator : null,
    possible_total_max: denominator ? (50 * upper) / denominator : null,
    resolved_test_count: all.filter((o) => ["pass", "fail"].includes(o)).length,
    applicable_test_count: all.length,
  };
}

File: data/evaluations/2026-09-05-codegen-pilot-01/deterministic-v1-aa75aca9abad1926b5f3/evaluator-source/src/evaluation/scoring.py
"""Pure scoring: no model calls, wall clock, randomness, or filesystem access."""
from fractions import Fraction

CATEGORIES = {
    'functional_requirements': 30,
    'connectivity': 30,
    'physical_constraints': 20,
    'deliverable_integrity': 20,
}
OUTCOMES = {'pass', 'fail', 'unknown', 'unsupported'}


def validate_rules(rules):
    if rules.get('version') != '1.0.0':
        raise ValueError('Unsupported rules version')
    if rules.get('weights') != CATEGORIES:
        raise ValueError('Rules must use the frozen 30/30/20/20 weights')
    applicability = rules.get('applicability', {})
    if set(applicability) != set(CATEGORIES) or any(type(x) is not bool for x in applicability.values()):
        raise ValueError('Every category needs explicit boolean applicability')
    seen = set()
    counts = dict.fromkeys(CATEGORIES, 0)
    for test in rules.get('tests', []):
        test_id = test.get('id')
        if not isinstance(test_id, str) or not test_id or test_id in seen:
            raise ValueError('Missing or duplicate test ID')
        seen.add(test_id)
        category = test.get('category')
        if category not in CATEGORIES or type(test.get('critical')) is not bool:
            raise ValueError('Invalid category or critical flag')
        if not isinstance(test.get('rule'), dict) or not test.get('requirement') or not test.get('source_criteria'):
            raise ValueError('Every test needs an executable rule and requirement provenance')
        counts[category] += 1
    if any(applicability[c] and counts[c] == 0 for c in CATEGORIES):
        raise ValueError('Applicable categories cannot have zero tests')
    return rules


def score(rules, outcomes):
    """Require exactly one evidence-backed decision per applicable rule."""
    validate_rules(rules)
    tests = {t['id']: t for t in rules['tests']}
    if len(outcomes) != len(tests) or {o['test_id'] for o in outcomes} != set(tests):
        raise ValueError('Missing, duplicate, or unexpected outcomes')
    for o in outcomes:
        if o.get('outcome') not in OUTCOMES:
            raise ValueError('Invalid outcome')
        if o['outcome'] in {'pass', 'fail'} and not o.get('evidence'):
            raise ValueError('Pass/fail requires evidence')
    by_id = {o['test_id']: o for o in outcomes}
    categories = []
    points = Fraction(0)
    lower = Fraction(0)
    upper = Fraction(0)
    denominator = 0
    any_unknown = False
    applicable_outcomes = []
    critical_failure = False
    for category, weight in CATEGORIES.items():
        applicable = rules['applicability'][category]
        selected = [by_id[t['id']]['outcome'] for t in tests.values() if t['category'] == category] if applicable else []
        n_pass = selected.count('pass')
        n_fail = selected.count('fail')
        n_unknown = sum(o in {'unknown', 'unsupported'} for o in selected)
        value = None
        lo = hi = None
        if applicable:
            denominator += weight
            applicable_outcomes.extend(selected)
            if n_unknown:
                any_unknown = True
                lo = 1 if n_pass else 0
                hi = 1 if n_fail else 2
            else:
                value = 2 if n_pass == len(selected) else 1 if n_pass else 0
                lo = hi = value
                points += Fraction(weight * value, 2)
            lower += Fraction(weight * lo, 2)
            upper += Fraction(weight * hi, 2)
            critical_failure |= any(t['critical'] and by_id[t['id']]['outcome'] == 'fail' for t in tests.values() if t['category'] == category)
        categories.append({'category': category, 'weight': weight, 'applicable': applicable, 'score': value,
                           'possible_score_min': lo, 'possible_score_max': hi,
                           'passed': n_pass, 'failed': n_fail, 'unknown_or_unsupported': n_unknown})
    total = float(100 * points / denominator) if denominator and not any_unknown else None
    overall = False if critical_failure else None if any_unknown or not denominator else all(o == 'pass' for o in applicable_outcomes)
    return {'category_scores': categories, 'total_score': total, 'overall_pass': overall,
            'critical_failure': critical_failure,
            'possible_total_min': float(100 * lower / denominator) if denominator else None,
            'possible_total_max': float(100 * upper / denominator) if denominator else None,
            'resolved_test_count': sum(o in {'pass', 'fail'} for o in applicable_outcomes),
            'applicable_test_count': len(applicable_outcomes)}
